.github/workflows/scripts/add-codeowners-to-pr.sh

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+#
+#   Copyright The OpenTelemetry Authors.
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# Adds code owners without write access as reviewers on a PR. Note that
+# the code owners must still be a member of the `open-telemetry`
+# organization.
+#
+# Note that since this script is considered a requirement for PRs,
+# it should never fail.
+
+set -euo pipefail
+
+if [[ -z "${PR:-}" ]]; then
+    echo "PR has not been set, please ensure it is set."
+    exit 0
+fi
+
+main () {
+    CUR_DIRECTORY=$(dirname "$0")
+    FILES=$(gh pr view "${PR}" --json files | jq -r '.files[].path')
+    COMPONENTS=$(grep -oE '^[a-z]+/[a-z/]+ ' < .github/CODEOWNERS)
+    REVIEWERS=""
+    LABELS=""
+    declare -A PROCESSED_COMPONENTS
+
+    for COMPONENT in ${COMPONENTS}; do
+        for FILE in ${FILES}; do
+            MATCH=$(echo "${FILE}" | grep -E "^${COMPONENT}" || true)
+
+            if [[ -z "${MATCH}" ]]; then
+                continue
+            fi
+
+            # If we match a file with a component we don't need to process the file again.
+            FILES=$(printf "${FILES}" | grep -v "${FILE}")
```

Comment:
If we loop through `FILES` as the outer loop I think we can avoid this extra manipulation.

Reply:
I played around with this a little, and the reason I want to loop through files in the inner loop and remove them from the list is that since the files will be in order, we can process all files for a component at once and shrink the list of files to check against the remaining components in the outer list. The same optimization wouldn't work if we processed components as the inner loop since there's a one-to-many relationship with components and files. Removing files from the inner loop like this did speed things up from my casual testing.

This whole process is inefficient due to the nested loops, but the only improvement I can think of would be using a trie, and I'd rather not implement that in bash. This still runs fairly quickly on my machine against some larger dependency PRs, so I don't think we would gain much from additional optimizations anyway.

If this sounds good to you, I'll document it in the script.

Reply:
> files will be in order

Great use of this fact, I wasn't aware that was happening.  I think what you've done is best.